client/driver/logging/universal_collector.go

```diff
@@ -76,69 +73,59 @@ func NewSyslogCollector(logger *log.Logger) *SyslogCollector {
 // LaunchCollector launches a new syslog server and starts writing log lines to
 // files and rotates them
 func (s *SyslogCollector) LaunchCollector(ctx *LogCollectorContext) (*SyslogCollectorState, error) {
-	addr, err := s.getFreePort(ctx.PortLowerBound, ctx.PortUpperBound)
+	l, err := s.getListener(ctx.PortLowerBound, ctx.PortUpperBound)
 	if err != nil {
 		return nil, err
 	}
-	s.logger.Printf("[DEBUG] sylog-server: launching syslog server on addr: %v", addr)
+	s.logger.Printf("[DEBUG] sylog-server: launching syslog server on addr: %v", l.Addr().String())
 	s.ctx = ctx
 	// configuring the task dir
 	if err := s.configureTaskDir(); err != nil {
 		return nil, err
 	}
 
-	channel := make(syslog.LogPartsChannel)
-	handler := syslog.NewChannelHandler(channel)
-
-	s.server = syslog.NewServer()
-	s.server.SetFormat(&CustomParser{logger: s.logger})
-	s.server.SetHandler(handler)
-	s.server.ListenTCP(addr.String())
-	if err := s.server.Boot(); err != nil {
-		return nil, err
-	}
+	channel := make(chan *SyslogMessage)
+	syslogServer := NewSyslogServer(l, channel, s.logger)
+	s.server = syslogServer
+	go syslogServer.Start()
 	logFileSize := int64(ctx.LogConfig.MaxFileSizeMB * 1024 * 1024)
 
-	ro, wo := io.Pipe()
-	lro, err := logrotator.NewLogRotator(filepath.Join(s.taskDir, allocdir.TaskLocal),
-		fmt.Sprintf("%v.stdout", ctx.TaskName), ctx.LogConfig.MaxFiles,
-		logFileSize, s.logger)
+	path := filepath.Join(s.taskDir, allocdir.TaskLocal)
+	lro, err := NewFileRotator(path, fmt.Sprintf("%v.stdout", ctx.TaskName),
+		ctx.LogConfig.MaxFiles, logFileSize, s.logger)
+
 	if err != nil {
 		return nil, err
 	}
 	s.lro = lro
-	go lro.Start(ro)
 
-	re, we := io.Pipe()
-	lre, err := logrotator.NewLogRotator(filepath.Join(s.taskDir, allocdir.TaskLocal),
-		fmt.Sprintf("%v.stderr", ctx.TaskName), ctx.LogConfig.MaxFiles,
-		logFileSize, s.logger)
+	lre, err := NewFileRotator(path, fmt.Sprintf("%v.stderr", ctx.TaskName),
+		ctx.LogConfig.MaxFiles, logFileSize, s.logger)
 	if err != nil {
 		return nil, err
 	}
 	s.lre = lre
-	go lre.Start(re)
 
-	go func(channel syslog.LogPartsChannel) {
+	go func(channel chan *SyslogMessage) {
 		for logParts := range channel {
 			// If the severity of the log line is err then we write to stderr
 			// otherwise all messages go to stdout
-			s := logParts["severity"].(Priority)
-			if s.Severity == s1.LOG_ERR {
-				we.Write(logParts["content"].([]byte))
+			if logParts.Severity == syslog.LOG_ERR {
+				s.lre.Write(logParts.Message)
+				s.lre.Write([]byte("\n"))
```

Comment:
Don't do two writes. Just append the new line character. Also `"\n"` is a string that you are forcing to be converted to a character. Use `'\n'` and below